File: operational-systems/ex_SystemCall_sys1.c
#include <unistd.h>     // Para definicoes como STDIN_FILENO
#include <sys/syscall.h> // Para constantes como SYS_read

#define STDIN_FD  0
#define STDOUT_FD 1

int main() {
    char buffer[256];
    long len;

    // syscall read: SYS_read(fd, buf, count)
    len = syscall(SYS_read, STDIN_FD, buffer, sizeof(buffer));  // <-- Leitura direta via syscall

    // Remover \n, se presente
    if (len > 0 && buffer[len - 1] == '\n') {
        len--;
    }

    // Inverter e imprimir com syscalls
    for (long i = len - 1; i >= 0; i--) {
        syscall(SYS_write, STDOUT_FD, &buffer[i], 1);           // <-- Escrita caractere por caractere via syscall
    }

    // Nova linha
    char newline = '\n';
    syscall(SYS_write, STDOUT_FD, &newline, 1);                 // <-- Escrita de nova linha via syscall

    return 0;
}

File: operational-systems/ex_SystemCall_sys1.py
import os

# Mensagem antes da leitura
os.write(1, b"Digite uma string: ")        # <-- SYS_write para stdout (fd 1)

# Leitura da entrada padrão (fd 0)
buffer = os.read(0, 256)                   # <-- SYS_read para stdin (fd 0)

# Verificar se a leitura foi bem-sucedida e remover '\n' do final
if buffer.endswith(b'\n'):
    buffer = buffer[:-1]

# Mensagem antes da exibição da string invertida
os.write(1, b"String invertida: ")         # <-- SYS_write para mensagem

# Impressão da string invertida, caractere por caractere
for i in reversed(buffer):
    os.write(1, bytes([i]))                # <-- SYS_write para cada byte

# Nova linha
os.write(1, b"\n")                         # <-- SYS_write para nova linha
